enterprise/reporting/error_reporting.go

```diff
@@ -0,0 +1,646 @@
+			&dbEdMetric.EDErrorDetails.ErrorMessage,
+			&dbEdMetric.EDConnectionDetails.DestType,
+		)
+		if err != nil {
+			edRep.log.Errorf("Failed while scanning rows(reported_at=%v): %v", queryMin.Int64, err)
+			return []*types.EDReportsDB{}, queryMin.Int64
+		}
+		metrics = append(metrics, dbEdMetric)
+	}
+	return metrics, queryMin.Int64
+}
+
+func (edRep *ErrorDetailReporter) aggregate(reports []*types.EDReportsDB) []*types.EDMetric {
+	groupedReports := lo.GroupBy(reports, func(report *types.EDReportsDB) string {
+		keys := []string{
+			report.WorkspaceID,
+			report.Namespace,
+			report.SourceDefinitionId,
+			report.SourceID,
+			report.DestinationID,
+			report.DestinationDefinitionId,
+			report.PU,
+			fmt.Sprint(report.ReportedAt),
+			report.DestType,
+		}
+		return strings.Join(keys, groupKeyDelimitter)
+	})
+	var edReportingMetrics []*types.EDMetric
+	groupKeys := lo.Keys(groupedReports)
+	sort.Strings(groupKeys)
+
+	for _, key := range groupKeys {
+		reports := groupedReports[key]
+		firstReport := reports[0]
+		edRepSchema := types.EDMetric{
+			EDInstanceDetails: types.EDInstanceDetails{
+				WorkspaceID: firstReport.WorkspaceID,
+				Namespace:   firstReport.Namespace,
+				InstanceID:  firstReport.InstanceID,
+			},
+			EDConnectionDetails: types.EDConnectionDetails{
+				DestinationID:           firstReport.DestinationID,
+				DestinationDefinitionId: firstReport.DestinationDefinitionId,
+				SourceID:                firstReport.SourceID,
+				SourceDefinitionId:      firstReport.SourceDefinitionId,
+				DestType:                firstReport.DestType,
+			},
+			PU: firstReport.PU,
+			ReportMetadata: types.ReportMetadata{
+				ReportedAt: firstReport.ReportedAt,
+			},
+		}
+		var errs []types.EDErrorDetails
+
+		reportsCountMap := lo.CountValuesBy(reports, func(rep *types.EDReportsDB) types.EDErrorDetails {
+			return types.EDErrorDetails{
+				StatusCode:   rep.StatusCode,
+				ErrorCode:    rep.ErrorCode,
+				ErrorMessage: rep.ErrorMessage,
+				EventType:    rep.EventType,
+			}
+		})
+
+		reportGrpKeys := lo.Keys(reportsCountMap)
+		sort.SliceStable(reportGrpKeys, func(i, j int) bool {
+			irep := reportGrpKeys[i]
+			jrep := reportGrpKeys[j]
+
+			return (irep.StatusCode < jrep.StatusCode ||
+				irep.ErrorMessage < jrep.ErrorMessage ||
+				irep.EventType < jrep.EventType)
+		})
+		for _, rep := range reportGrpKeys {
+			errs = append(errs, types.EDErrorDetails{
+				StatusCode:   rep.StatusCode,
+				ErrorCode:    rep.ErrorCode,
+				ErrorMessage: rep.ErrorMessage,
+				EventType:    rep.EventType,
+				Count:        reportsCountMap[rep],
+			})
+		}
+		edRepSchema.Errors = errs
+		edReportingMetrics = append(edReportingMetrics, &edRepSchema)
+	}
+	return edReportingMetrics
+}
+
+func (edRep *ErrorDetailReporter) sendMetric(ctx context.Context, clientName string, metric *types.EDMetric) error {
+	payload, err := json.Marshal(metric)
+	if err != nil {
+		return fmt.Errorf("marshal failure: %v", err)
```

Comment:
We can use error wrapping here:

```suggestion
		return fmt.Errorf("marshal failure: %w", err)
```